db/data/20210804154421_backfill_benefitting_countries.rb

```diff
@@ -0,0 +1,51 @@
+# Run me with `rails runner db/data/20210804154421_backfill_benefitting_countries.rb`
+
+benefitting_countries_hash = Codelist.new(type: "intended_beneficiaries").list
+region_to_subregion_hash = Codelist.new(type: "region_to_subregion_mapping", source: "beis").list
+activities_to_update = Activity.where.not(geography: nil).where(benefitting_countries: nil)
+total_activities_to_update = activities_to_update.count
+
+activities_to_update.each_with_index do |activity, counter|
```

Comment:
I wonder if this might be quicker if we updated all the recipient countries first, then looped through the regions in turn and did a bulk update on each set of activities with a given region?

Reply:
Something like this:

```
recipient_country_activities = Activity.where(geography: "recipient_country")

recipient_country_activities.each do |activity|
    recipient_country = activity.recipient_country
    intended_beneficiaries = activity.intended_beneficiaries || []

    activity.benefitting_countries = ([recipient_country] + intended_beneficiaries).compact.uniq
    activity.save(validate: false)
end

low_level_region_activities = Activity.where(geography: "recipient_region", recipient_region: benefitting_countries_hash.keys).group_by(&:recipient_region)

benefitting_countries_hash.each do |region, countries|
   next if low_level_region_activities[region].blank?

   activities = Activity.where(id: low_level_region_activities[region].pluck(:id))
   activities.update_all(benefitting_countries: countries.map { |c| c["code"] })
end

other_region_activites =  Activity.where(geography: "recipient_region", recipient_region: region_to_subregion_hash.keys).group_by(&:recipient_region)

region_to_subregion_hash.each do |region, subregions|
   next if other_region_activites[region].blank?

  activities = Activity.where(id: other_region_activites[region].pluck(:id))
  countries = subregions.map { |region| benefitting_countries_hash.fetch(region, {}) }.flatten
   activities.update_all(benefitting_countries: countries.map { |c| c["code"] }.reject(&:blank?))
end

```

Reply:
Please review the update, [in light of this thread](https://dxw.slack.com/archives/CN0H2L066/p1628602900114600?thread_ts=1628600547.107200&cid=CN0H2L066) 😅

Reply:
Just made a tweak that should do it

Reply:
What's the tweak?

Reply:
Just matching against the presence (or otherwise) of `recipient_country` / `recipient_region` and removing the specific check for geography

Reply:
Comments should have git history applied to them 😆

Reply:
Oh I see. That doesn't solve the issue that my latest commit was trying to address, which is to add up *all* the data we find, including when an activity has *both* recipient country and region.

If we set `benefitting_countries` when using the region, we overwrite whatever came from the recipient country and intended beneficiaries, which might not be perfectly included in the region.

Reply:
Oh course - that's annoying. I'll give it a bit more thought after lunch